### ISSUE TITLE
[Question] How to use a HttpGateway in a ASP .NET Web API project

### ISSUE THREAD

**Issue by MegaMax93** [c3-ls/ServiceFabric-Http#13]
_If this is the wrong place for asking questions, I am sorry. Remove this issue if it's the case._ 

I'm sorry, but it's not yet clear for me how I should use your HttpServiceGateway correctly.
I have created a new ASP .NET 5 Web API project into my service fabric solution and added the nuget references HttpCommunication and HttpServiceGateway (0.3.0). 
I want to have one HttpGateway microservice, which is resolving the endpoints of my other two (in a local service fabric cluster deployed) microservices (MasterDataMService and ActorTestService).

So I added this code to my `HttpServiceGateway.Startup.cs`:

``` csharp
public class Startup
{
  public Startup(IHostingEnvironment env)
  {
    var builder = new ConfigurationBuilder()
          .AddJsonFile("appsettings.json")
          .AddEnvironmentVariables();
    Configuration = builder.Build();
  }

  public void ConfigureServices(IServiceCollection services)
  {
    services.AddMvc();
    services.Configure<HttpCommunicationOptions>(this.Configuration.GetSection("HttpCommunication"));
    services.AddServiceFabricHttpCommunication();
  }

  public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
  {
      loggerFactory.AddConsole(Configuration.GetSection("Logging"));
      loggerFactory.AddDebug();
      app.UseIISPlatformHandler();
      app.UseStaticFiles();
      app.UseMvc();  // do I have to remove that?
      this.ConfigureHttpServiceGateways(app);
  }

  private void ConfigureHttpServiceGateways(IApplicationBuilder app)
  {
    app.RunHttpServiceGateway("/OrganizationMasterDataMService", "fabric:/DataServiceFabric/OrganizationMasterDataMService");
    app.RunHttpServiceGateway("/ActorTestService", "fabric:/DataServiceFabric/ActorTestService");
  }

  public static void Main(string[] args) => WebApplication.Run<Startup>(args);
}
```

How do I call my MasterDataMService in my controller now? I added this code to my `HttpServiceGateway.ValuesController.cs`:

``` csharp
[Route("api/[controller]")]
public class ValuesController : Controller
{
  private readonly IHttpCommunicationClientFactory ClientFactory = new HttpCommunicationClientFactory(new LoggerFactory(), ServicePartitionResolver.GetDefault(), new OptionsManager<HttpCommunicationOptions>(null));
  private readonly Uri MasterDataMServiceUri = new Uri("fabric:/DataServiceFabric/MasterMDataService");

  [HttpGet("{query}")]
  public async Task<string> Get(string query)
  {
    var client = new ServicePartitionClient<HttpCommunicationClient>(this.ClientFactory, this.MasterDataMServiceUri);

    var result = new HttpResonseMessage();
    await client.InvokeWithRetry(async x =>
    {
      HttpRequestMessage req = new HttpRequestMessage(HttpMethod.Get, "api/values"); // how do I have to specifiy the second parameter?

      HttpResponseMessage response = await x.HttpClient.SendAsync(req, this.HttpContext.RequestAborted);

      await this.HttpContext.Response.WriteAsync(DateTime.Now + " - Result from API: ");
      await this.HttpContext.Response.WriteAsync("Status: " + response.StatusCode + "; Body: ");
      await response.Content.CopyToAsync(this.HttpContext.Response.Body);

      result = response;
    });
    return result.Content.ReadAsStringAsync();
```

By executing this code in ran into this error:

```
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Microsoft.AspNet.Hosting.Internal.HostingEngine: Information: Request finished in 4,2531ms 500
Microsoft.AspNet.Server.WebListener.MessagePump: Error: ProcessRequestAsync
System.InvalidOperationException: The endpoint address is not valid. Please resolve again.
   at C3.ServiceFabric.HttpCommunication.HttpCommunicationClientFactory.CreateEndpointUri(String endpoint)
   at C3.ServiceFabric.HttpCommunication.HttpCommunicationClientFactory.CreateClientAsync(String endpoint, CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Communication.Client.CommunicationClientFactoryBase`1.<CreateClientWithRetriesAsync>d__1e.MoveNext()
--- End of stack trace from the previous location where the exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ServiceFabric.Services.Communication.Client.CommunicationClientFactoryBase`1.<GetClientAsync>d__f.MoveNext()
```

What do I also have to do for setting the communication to my MasterDataMService up correctly?

Thanks for your time.

**Comment by cwe1ss** [comment]
Hi. 

> app.UseMvc();  // do I have to remove that?

You don't have to use MVC in your gateway. You would only need it, if you want to have additional controllers/views in your app. Same for `app.UseStaticFiles()`. The gateway component is a middleware and is not dependent on MVC.

If you have a gateway, you don't need `C3.ServiceFabric.HttpCommunication` in your calling application. You can just do a regular HTTP request to the load balancer URL of your gateway. In other words, your calling application has no idea that you are using Service Fabric behind the gateway. 

Your gateway must have a fixed port in your Service Fabric cluster - e.g. 5010 (ServiceManifest.xml -> Endpoints):

``` xml
<Endpoints>
  <Endpoint Name="HttpGatewayTypeEndpoint" Protocol="http" Type="Input" Port="5010" />
</Endpoints>
```

You will then have to setup a rule in your load balancer for this port. You can e.g. say that the public URL 80 should map to the internal port 5010.

This way, you should now be able to call your gateway on your public load balancer address - e.g. http://your-cluster.westeurope.cloudapp.azure.com/OrganizationMasterDataMService

As soon as this works, you can do regular HttpClient calls to this url in your calling application.

`C3.ServiceFabric.HttpCommunication` is only necessary, if you want to do HTTP communication between your services within your cluster WITHOUT using a gateway. 

Note that this example means, that your gateway will be publically available. If you don't want that you either have to protect your gateway or use an internal load balancer. You probably also should use HTTPS instead of HTTP.

Hope this helps?!?

**Comment by cwe1ss** [comment]
oh, I think I misunderstood something.

> How do I call my MasterDataMService in my controller now? I added this code to my HttpServiceGateway.ValuesController.cs:

You would not need this controller at all. the gateway itself is a middleware and setting it up in Startup.cs is all you need to do. 

``` csharp
app.RunHttpServiceGateway("/OrganizationMasterDataMService", "fabric:/DataServiceFabric/OrganizationMasterDataMService");
```

This line makes sure that every call to http://url-to-gateway/OrganizationMasterDataMService/whatever gets forwarded to your "fabric:/DataServiceFabric/OrganizationMasterDataMService".

have a look at the samples folder. If you deploy the GatewaySample into your local cluster, you will have two services http://localhost:assigned-port for the "HttpService" (check SF cluster manager for the actual port) and http://localhost:5402 for the gateway (5402 is a fixed port assigned in ServiceManifest.xml). You should be able to call http://localhost:assigned-port/api/values and http://localhost:5402/service1/api/values and it should give you the same results.

**Comment by MegaMax93** [comment]
Hi Christian,

thanks for your quick reply. 

I think I have understood now how to use your `HttpServiceGateway`. Thanks for all the tips you gave me.
I gave my `GatewayService`the port 5010 in the `ServiceManifest` and removed the unnecessary `app.UseMvc()` and `app.UseStaticFiles()`.

But at the moment I am still stuck. When I call `http://localhost:5010/OrganizationMasterDataMService/api/values/query/somequery` in the browser I get an `InvalidOperationException` while debugging:

```
Microsoft.AspNet.Hosting.Internal.HostingEngine: Information: Request starting HTTP/1.1 GET http://localhost:5010/OrganizationMasterDataMService/api/values/query/somequery
C3.ServiceFabric.HttpCommunication: Error: Service call failed. (The endpoint address is not valid. Please resolve again.)
System.InvalidOperationException: The endpoint address is not valid. Please resolve again.
   at C3.ServiceFabric.HttpCommunication.HttpCommunicationClientFactory.CreateEndpointUri(String endpoint)
   at C3.ServiceFabric.HttpCommunication.HttpCommunicationClientFactory.CreateClientAsync(String endpoint, CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Communication.Client.CommunicationClientFactoryBase`1.<CreateClientWithRetriesAsync>d__1e.MoveNext()
Microsoft.AspNet.Hosting.Internal.HostingEngine: Information: Request finished in 0ms 500
Microsoft.AspNet.Server.WebListener.MessagePump: Error: ProcessRequestAsync
System.InvalidOperationException: The endpoint address is not valid. Please resolve again.
   at C3.ServiceFabric.HttpCommunication.HttpCommunicationClientFactory.CreateEndpointUri(String endpoint)
   at C3.ServiceFabric.HttpCommunication.HttpCommunicationClientFactory.CreateClientAsync(String endpoint, CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Communication.Client.CommunicationClientFactoryBase`1.<CreateClientWithRetriesAsync>d__1e.MoveNext()
--- End of stack trace from the previous location where the exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ServiceFabric.Services.Communication.Client.CommunicationClientFactoryBase`1.<GetClientAsync>d__f.MoveNext()
--- End of stack trace from the previous location where the exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ServiceFabric.Services.Communication.Client.CommunicationClientFactoryBase`1.<GetClientAsync>d__0.MoveNext()
--- End of stack trace from the previous location where the exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ServiceFabric.Services.Communication.Client.ServicePartitionClient`1.<InvokeWithRetryAsync>d__7`1.MoveNext()
```

Do I have to configure something else? I have two endpoints in my OrganizationMasterDataMService. One for providing my WebAPI (WebApiEndpoint) and the other for communicating over the `ServiceProxy` (ServiceEndpoint). Could this be a problem? 

**OrganizationMasterDataMService.ServiceManifest:**

``` xml

<?xml version="1.0" encoding="utf-8"?>
<ServiceManifest Name="OrganizationMasterDataMServicePkg"
                 Version="1.0.0"
                 xmlns="http://schemas.microsoft.com/2011/01/fabric"
                 xmlns:xsd="http://www.w3.org/2001/XMLSchema"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <ServiceTypes>
    <!-- This is the name of your ServiceType. 
         This name must match the string used in RegisterServiceType call in Program.cs. -->
    <StatelessServiceType ServiceTypeName="OrganizationMasterDataMServiceType" />
  </ServiceTypes>

  <!-- Code package is your service executable. -->
  <CodePackage Name="Code" Version="1.0.0">
    <EntryPoint>
      <ExeHost>
        <Program>OrganizationMasterDataMService.exe</Program>
      </ExeHost>
    </EntryPoint>
  </CodePackage>

  <!-- Config package is the contents of the Config directoy under PackageRoot that contains an 
       independently-updateable and versioned set of custom configuration settings for your service. -->
  <ConfigPackage Name="Config" Version="1.0.0" />

  <Resources>
    <Endpoints>
      <!-- This endpoint is used by the communication listener to obtain the port on which to 
           listen. Please note that if your service is partitioned, this port is shared with 
           replicas of different partitions that are placed in your code. -->
      <Endpoint Name="WebApiEndpoint" Type="Input" Protocol="http" Port="80"/>
      <Endpoint Name="ServiceEndpoint"/>
      <Endpoint Name="ReplicatorEndpoint"/>
    </Endpoints>
  </Resources>
</ServiceManifest>
```

Or is there something else I have to check?

Thanks for your help and kind regards,
Max

**Comment by cwe1ss** [comment]
A few more questions:
- Since you use port 80 for the service, can you successfully call your service on http://localhost/ ?
- Have you verified that the service fabric URL fabric:/DataServiceFabric/OrganizationMasterDataMService is correct?
- Which addresses does Service Fabric Explorer show for this service? To see that, navigate to the most detailed entry in the left navigation (the actual instance). Does it show other addresses than just http://localhost:80 or http://your-ip:80 ?

I have not yet tried it with services that have more than one endpoint. Maybe I have to check the protocol if there are more than one.

**Comment by MegaMax93** [comment]
- Yes, I can successfull call my service on `localhost/OrganizationMasterDataMService/api/values/query/somequery`
- I stepped several times through my sorcecode and I can verify that the service fabric URL `fabric:/DataServiceFabric/OrganizationMasterDataMService` is equal on every spot.
- It shows me these addresses:

```
{"Endpoints":
{"RemotingListener":"net.tcp:\/\/localhost:32002\/aa76e0b0-0c34-4e76-adc4-8428a1ac2d6a\/37c913cd-73ee-4a11-aaec-f8a7643b7d79-131039889137819658",
"OwinListener":"http:\/\/localhost:80\/OrganizationMasterDataMService\/"}}
```

**Comment by cwe1ss** [comment]
Sorry for the late reply. I'm afraid having a service with multiple endpoint types is not supported right now. 
Unfortunately, I don't have time during the next days to look at this. It's also a bit tough at the moment because of all the dnx/dotnet cli/netstandard/... changes in .NET Core. Also, the master branch already targets an ASP.NET Core RC2 daily build (20100) and I wanted to wait until these things finalize before I release a new version.

If you need this fixed before I have time, you could try to resolve it yourself. You would have to create a branch from the v0.3.0 tag. I'm pretty sure the problem is in `HttpCommunicationClient.CreateClientAsync(string endpoint, CancellationToken cancellationToken)`. Maybe this `endpoint` string will contain the json object from your previous post, so we just have to choose one that starts with `http:` or `https:`.

**Comment by MegaMax93** [comment]
Hi Christian,

no problem. Thanks for taking the time to help me.
Yeah, I read about all the changes. I updated to ServiceFabric GA (2.0.135) and had to realize that the ASP.NET Web API templates are gone. But they are coming back in RC2. I tried to add a new ASP.NET project "manually", but I was not able to deploy my service fabric application by this approach. So it seems it might be the best solution to wait until the support for RC2 is released in service fabric. 

At the moment it's not necessary to have this fixed soon. For my master's degree I have to develop a microservice-architecture approach with service fabric. So I'm trying out some fundamental stuff and a GatewayService would be nice to have.
I cloned your repository to my local dev machine and did some debugging, but I was not able to fix it. In the `HttpCommunicationClient.CreateClientAsync(string endpoint, CancellationToken cancellationToken)` method the `endpoint` attribute has only one  TCP-endpoint:

`net.tcp://localhost:31002/660e1f51-5202-4107-800e-28375840061c/1dff9486-91f2-4f1a-a548-b597470a4d5f-131051015989251379`

So it seems the problem is somewhere before. But I don't know where. And I have verified, that the OrganizationMasterDataMService has two endpoints.

After the debugging session I removed the "RemotingListener" (TCP) endpoint of my OrganizationMasterDataMService, so that I have only the "OwinListener" (HTTP) active. I also changed the port of the microservice (5020) for testing.
I verified that `http://localhost:5020/OrganizationMasterDataMService/api/values/query/somequery` is working. After that I tried to access it by the GatewayService on Port 5010: `http://localhost:5010/OrganizationMasterDataMService/api/values/query/somequery`. But I get an "Service Unavailable: HTTP Error 503. The service is unavailable"

And here the stacktrace:

```
Microsoft.AspNet.Hosting.Internal.HostingEngine: Information: Request starting HTTP/1.1 GET http://localhost:5010/OrganizationMasterDataMService/api/values/query/somequery  
C3.ServiceFabric.HttpCommunication: Warning: Retrying Service call. Reason: HTTP 503, Details: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Service Unavailable</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Service Unavailable</h2>
<hr><p>HTTP Error 503. The service is unavailable.</p>
</BODY></HTML>

C3.ServiceFabric.HttpCommunication: Warning: Retrying Service call. Reason: HTTP 503, Details: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Service Unavailable</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Service Unavailable</h2>
<hr><p>HTTP Error 503. The service is unavailable.</p>
</BODY></HTML>

C3.ServiceFabric.HttpServiceGateway: Warning: Service returned non retryable error. Reason: HTTP ServiceUnavailable
Microsoft.AspNet.Hosting.Internal.HostingEngine: Information: Request finished in 1,8734ms 503 text/html; charset=us-ascii
```

Do I have configured something wrong?

I added a freshly new microservice (WebGatewayTest) with an OwinListener and added the URI Mapping in the `Startup.cs`. This worked. When I access `localhost:5010/WebGatewayTest/api/values` I get some lovely "value1, value2".

**Comment by cwe1ss** [comment]
thx for sharing the error details. It's interesting that the HTTP endpoint doesn't show up.

The only thing I noticed is that your gateway-url might not be correct. If you can access your service directly at `http://localhost:5020/OrganizationMasterDataMService/api/values` then your gateway url would need the additional path-part for the configured service name. So if you have configured it with RunHttpServiceGateway("/**SERVICE**", "fabric:/..."), then the gateway url should be http://localhost:5010/**SERVICE**/OrganizationMasterDataMService/api/values

PS: Note that there's also a more or less official draft of a gateway from Microsoft here: https://github.com/weidazhao/Hosting - it's very similar and the project contains more details about how to get it running with dotnet cli. Maybe that helps!

**Comment by cwe1ss** [comment]
the new version of the Service Fabric SDK provides a way to pass a "ListenerName". I will implement this as an option for the gateway, so I hope this will resolve your issue.

**Comment by cwe1ss** [comment]
This is now in the master branch. It would be great if you could test this with your service. You should have to change the code in your gateway Startup.cs to something like this:

``` csharp
app.RunHttpServiceGateway("/service", new HttpServiceGatewayOptions
{
   ServiceName = new Uri("fabric:/DataServiceFabric/OrganizationMasterDataMService"),
   ListenerName = "OwinListener"
});
```

If you need this as a nuget package, please add `https://www.myget.org/F/c3-oss/api/v3/index.json` to your nuget.config to get the package from our dev feed. 
I want to wait a few more days until I release the packages to nuget.org because I'm not yet sure if I should target 1.0.0-RC2-\* or 1.0.0-\* in my ASP.NET Core packages. (there are some RC3 packages in the ASP.NET Core feeds).

**Comment by MegaMax93** [comment]
Hi Christian,
sorry for the late reply, I was very busy in the last few days.
You were right, my url was wrong. `http://localhost:5010/OrganizationMasterDataMService/OrganizationMasterDataMService/api/values` worked for me. Thanks! :)

After that I wanted to test your new ListenerName functionality. Sadly I was not able to successfully update your nuget package to 0.4.0-rc2-57 from your dev feed.

I ran into these errors:

```
Unable to locate Dependency Microsoft.Extensions.Logging.Abstractions >= 1.0.0-rc3-20588
Unable to locate Dependency Microsoft.Extensions.Options >= 1.0.0-rc3-20588
Unable to locate Dependency Microsoft.AspNetCore.Http.Abstractions >= 1.0.0-rc3-20588
Writing lock file D:\DataService\DataServiceFabric\Web1\project.lock.json
Restore complete, 1365ms elapsed
Errors in D:\DataService\DataServiceFabric\Web1\project.json
    Unable to locate Dependency Microsoft.Extensions.Logging.Abstractions >= 1.0.0-rc3-20588
    Unable to locate Dependency Microsoft.Extensions.Options >= 1.0.0-rc3-20588
    Unable to locate Dependency Microsoft.AspNetCore.Http.Abstractions >= 1.0.0-rc3-20588

NU1001 The dependency Microsoft.AspNetCore.Http.Abstractions >= 1.0.0-rc3-20588 could not be resolved.
NU1001 The dependency Microsoft.Extensions.Options >= 1.0.0-rc3-20588 could not be resolved.
```

The "Web1" service is my current gateway service. I created this ASP.NET 5 project before service fabric GA was released, so might I am referencing an old nugets? Or do I have to create a new ASP.NET 5 project (or other) especially for service fabric GA? Sorry for the dumb question, but I am a little bit overwhelmed with all the changes that recently ocurred...

Here is my **project.json:**

``` json
{
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },

  "dependencies": {
    "C3.ServiceFabric.HttpServiceGateway": "0.4.0-rc2-57",
    "Microsoft.AspNet.Diagnostics": "1.0.0-rc1-final",
    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-rc1-final",
    "Microsoft.AspNet.Mvc": "6.0.0-rc1-final",
    "Microsoft.AspNet.Server.Kestrel": "1.0.0-rc1-final",
    "Microsoft.AspNet.StaticFiles": "1.0.0-rc1-final",
    "Microsoft.AspNet.WebApi": "5.2.3",
    "Microsoft.Extensions.Configuration.FileProviderExtensions": "1.0.0-rc1-final",
    "Microsoft.Extensions.Configuration.Json": "1.0.0-rc1-final",
    "Microsoft.Extensions.Logging": "1.0.0-rc1-final",
    "Microsoft.Extensions.Logging.Console": "1.0.0-rc1-final",
    "Microsoft.Extensions.Logging.Debug": "1.0.0-rc1-final",
    "Microsoft.Extensions.OptionsModel": "1.0.0-rc1-final",
    "Microsoft.Owin": "3.0.1",
    "Microsoft.ServiceFabric.AspNet.Hosting": "1.0.0-rc1",
    "MongoDB.Bson": "2.2.3",
    "MongoDB.Driver": "2.2.3"
  },

    "commands": {
        "web": "Microsoft.AspNet.Server.Kestrel"
    },

    "frameworks": {
      "dnx451": {
        "dependencies": {
          "OrganizationMasterDataMService.Interfaces": "1.0.0-*",
          "System.Web.Mvc": "1.0.0-*"
        },
        "frameworkAssemblies": {
          "System.Web": "4.0.0.0"
        }
      }
    },

    "exclude": [
        "wwwroot",
        "node_modules",
        "PackageRoot"
    ],
    "publishExclude": [
        "**.user",
        "**.vspscc",
        "PackageRoot"
    ]
}
```

I also updated my local repository of ServiceFabric-Http to build it by myself, but there I ran into some DNX issues. Visual Studio reports that I have to install dnx-clr-win-x86.1.0.0-rc2-20222 but neither Visual Studio or `dnvm install 1.0.0-rc2-20222 -a x86 -r clr -OS win` and `dnvm install 1.0.0-rc2-20222 -a x86 -r clr -OS win -u` commands in the dnvm console application could  successfully install it. -> "There are no runtimes matching the name dnx-clr-win-x86 on feed https://www.nuget.org/api/v2." I am confuesd. This whole dnx thing is very annyoing.

**Comment by cwe1ss** [comment]
Unfortunately, it's really not the best time to play with ASP.NET Core right now. If you don't actually need it, I'd advice you to wait until RC2 is released. We have to use it and we have to change stuff almost daily :(

there are quite a few issues in your post: :)

rc2-20222 is not yet on the "aspnetvnext" feed. to install it with dnvm install, you have to set an environment variable called "DNX_UNSTABLE_FEED" that points to https://www.myget.org/F/aspnetcirelease/api/v3/index.json (you still have to add the "-u" parameter)

your project is still targeting RC1 - you can't integrate C3.ServiceFabric.HttpServiceGateway there because it is compiled against RC2 daily builds. You would have to update it first.

Since the gateway will usually experience high load (it has to process every request), there really shouldn't be anything else in this application. So I would not combine the gateway with anything else.

your project references "System.Web" and "System.Web.Mvc" which is weird. Both are from the pre ASP.NET Core times and should not be required!

hope that helps.

**Comment by MegaMax93** [comment]
Thanks for the hints and tips you gave me. I have created this project when I started learning service fabric and tried out some things in there and I didn't cleaned it up afterwards. So that is why I have some weird references. It was the only working service in my service fabric cluster. And all other ASP.NET projects that I have freshly new created didn't worked because of some strange and frustrating "Partition is below target replica or instance count" errors. So I had to stick with my old project.

I have set the environment variable but when I try to install it with `dnvm install 1.0.0-rc2-20222 -a x86 -r clr -OS win -u`, I get an **"Unable to find any runtime packages on the feed!"**. Some people report about proxy issues that are causing this error message, but I don't use a proxy.

Is there a simple way for migrating a RC1 project to RC2? I was not able to successfully update it to RC2. Or is there an other way to create a freshly new RC2 project and test it with that?

I ran from one error into the next error what is at really frustrating and time consuming. I also don't want to annoy you with all my issues.. :(

**Comment by MegaMax93** [comment]
Hi Christian,

after Microsoft has finally released rc2 I was able to successfully set up a HttpGatewayService into my local cluster. I also tested the "ListenerName" feature, that works fine. Thanks!

In the last few days I successfully secured my WebAPI service and use HTTPS for the communication now. So I want HTTPS for my HttpGateway service too. But I was not able to reach my HTTPS MasterDataService over the  HttpGateway service. My browser ended up waiting for a response from the server. Unfortunately Visual Studio didn't provide me with any helpful stack strace and I dont know how to do a reasonable debugging here. I have created another Issue (#26) for this, if this is okay.
